jvm/src/main/java/com/microsoft/bond/protocol/CompactBinaryReader.java

```diff
@@ -0,0 +1,331 @@
+package com.microsoft.bond.protocol;
+
+import com.microsoft.bond.BondDataType;
+import com.microsoft.bond.InvalidBondDataException;
+import com.microsoft.bond.Metadata;
+import com.microsoft.bond.ProtocolType;
+import com.sun.deploy.util.StringUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.LinkedList;
+
+/**
+ * Implements Compact Binary deserialization.
+ * Refer to {@link https://microsoft.github.io/bond/reference/cpp/compact__binary_8h_source.html} for details.
+ */
+public final class CompactBinaryReader implements TaggedProtocolReader {
+
+    private final BinaryStreamReader reader;
+    private final short version;
+
+    public CompactBinaryReader(InputStream inputStream, short version) {
+        // check protocol version
+        if (version != 1 && version != 2)
+            throw new IllegalArgumentException("Invalid protocol version: " + version);
+
+        this.reader = new BinaryStreamReader(inputStream);
+        this.version = version;
+    }
+
+    private BondDataType readType() throws IOException {
+        return BondDataType.get(this.reader.readInt8());
+    }
+
+    @Override
+    public void readStructBegin() throws IOException {
+        this.reader.readVarUInt32();
```

Comment:
Yes. Also, my unit testing revealed a bug here for V1, it should be done only for V2. Fixed in next iteration.

---
In reply to: [119447363](https://github.com/Microsoft/bond/pull/465#discussion_r119447363) [](ancestors = 119447363)